nemo/collections/nlp/models/language_modeling/megatron_t5_prompt_learning_model.py

```diff
@@ -0,0 +1,350 @@
+from nemo.collections.nlp.parts.utils_funcs import get_last_rank
+from nemo.utils import logging
+
+try:
+    from apex.transformer import parallel_state
+
+    HAVE_APEX = True
+
+except (ImportError, ModuleNotFoundError):
+    HAVE_APEX = False
+
+
+__all__ = ['MegatronT5PromptLearningModel']
+
+
+class MegatronT5PromptLearningModel(MegatronBasePromptLearningModel):
+    """
+    Model class for prompt-tuning or p-tuning a pretrained Megatron T5 model. 
+
+    Prompt Tuning initalizes virtual prompt embeddings directly from a copy of
+    certain token embeddings from the the pretrained T5 model's vocabulary
+    and directly tunes these embedding weights. The token embeddings used in 
+    initalization are specified by the user in the config file. The model can 
+    be prompt-tuned for multiple tasks at once. Virtual prompts are stored in a 
+    prompt table and can be added or deleted without disrupting virtual prompts 
+    for other tasks. 
+
+    P-tuning initializes an LSTM encoder model that generates virtual prompt
+    embeddings for every task. Each task shares the same encoder. After p-tuning
+    is compelete, the learned virtual prompts can be saved to the prompt table
+    using add_ptuned_prompts_to_prompt_table(). Thus, if a user wants to add a 
+    new virtual prompt via p-tuning, they do not need to retrain on all previous 
+    tasks. This gives p-tuning the same task flexiblity as prompt-tuning.
+    """
+
+    def __init__(self, cfg: DictConfig, trainer: Trainer):
+        super().__init__(cfg, trainer)
+
+    def forward(
+        self, input_ids, dec_input, enc_mask, dec_mask, position_ids, taskname_ids, labels=None, inference=False,
+    ):
+        """
+        Special forward method for p-tuning/prompt-tuning pretrained
+        T5 style models.
+        """
+        # Get embeddings for text tokens and insert virtual token embeddings
+        if inference:
+            input_embeds = self.embed_input_inference(input_ids, taskname_ids)
+        else:
+            input_embeds = self.embed_input_train(input_ids, taskname_ids)
+
+        position_embeddings = self.frozen_model.enc_dec_model.encoder_embedding.position_embeddings(position_ids)
+        encoder_input = input_embeds + position_embeddings
+
+        # Call forward on T5 model with preprocessed embeddings
+        if self.float_type == torch.float32:
+            output = self.frozen_model.enc_dec_model(
+                enc_input_ids=None,
+                enc_attn_mask=enc_mask,
+                dec_input_ids=dec_input,
+                dec_attn_mask=dec_mask,
+                token_type_ids=None,
+                labels=labels,
+                enc_hidden_states=None,
+                output_enc_hidden_only=False,
+                enc_input=encoder_input,
+            )
+        else:
+            with torch.autocast(device_type="cuda", dtype=self.float_type):
+                output = self.frozen_model.enc_dec_model(
```

Comment:
You shouldn't call the parent model's forward function since this will not work with pipeline parallel and will need significant refactoring to work with it. Please write another method that uses apex's `forward_backward_pipelining_without_interleaving` or `forward_backward_no_pipelining`.